unit_tests/request_client_server/test_main.cpp

```diff
@@ -630,44 +647,104 @@ TEST_CASE ("Two request_sync_client_server communicate with each other", "")
 
   test_two_request_sync_client_server_env env;
 
-  constexpr int MESSAGE_COUNT = 4200;
-
-#define BIND_ENV_FUNC(foo) \
-  std::bind (&(test_two_request_sync_client_server_env::foo), std::ref (env), std::placeholders::_1, std::placeholders::_2)
+  SECTION ("Push and send requests and confirm they are handled")
+  {
+    constexpr int MESSAGE_COUNT = 4200;
 
 #define REPEAT_REQUEST_WITH_MSGID(req_foo, msgid) \
   for (int i = 0; i < MESSAGE_COUNT; ++i) req_foo (msgid, i)
 
-  std::vector<std::thread> threads;
-  threads.emplace_back ([&] ()
-  {
-    REPEAT_REQUEST_WITH_MSGID (env.send_recv_request_on_scs_one, reqids_1_to_2::_0);
-  });
-  threads.emplace_back ([&] ()
-  {
-    REPEAT_REQUEST_WITH_MSGID (env.push_request_on_scs_one, reqids_1_to_2::_1);
-  });
-  threads.emplace_back ([&] ()
-  {
-    REPEAT_REQUEST_WITH_MSGID (env.send_recv_request_on_scs_two, reqids_2_to_1::_0);
-  });
-  threads.emplace_back ([&] ()
-  {
-    REPEAT_REQUEST_WITH_MSGID (env.push_request_on_scs_two, reqids_2_to_1::_1);
-  });
-  threads.emplace_back ([&] ()
+    std::vector<std::thread> threads;
+    threads.emplace_back ([&] ()
+    {
+      REPEAT_REQUEST_WITH_MSGID (env.send_recv_request_on_scs_one, reqids_1_to_2::_0);
+    });
+    threads.emplace_back ([&] ()
+    {
+      REPEAT_REQUEST_WITH_MSGID (env.push_request_on_scs_one, reqids_1_to_2::_1);
+    });
+    threads.emplace_back ([&] ()
+    {
+      REPEAT_REQUEST_WITH_MSGID (env.send_recv_request_on_scs_two, reqids_2_to_1::_0);
+    });
+    threads.emplace_back ([&] ()
+    {
+      REPEAT_REQUEST_WITH_MSGID (env.push_request_on_scs_two, reqids_2_to_1::_1);
+    });
+    threads.emplace_back ([&] ()
+    {
+      REPEAT_REQUEST_WITH_MSGID (env.push_request_on_scs_two, reqids_2_to_1::_1);
+    });
+
+    for (auto &th : threads)
+      {
+	th.join ();
+      }
+
+    env.wait_for_all_messages ();
+
+    require_all_sent_requests_are_handled ();
+  }
+
+  SECTION ("Detect a failure while push request")
   {
-    REPEAT_REQUEST_WITH_MSGID (env.push_request_on_scs_two, reqids_2_to_1::_1);
-  });
+    {
+      // Ensure no failure is detected.
+      env.push_request_on_scs_one (reqids_1_to_2::_1, 0);
+      env.wait_for_all_messages ();
 
-  for (auto &th : threads)
+      const auto [send_error, recv_error] = env.get_request_errors_on_scs_one ();
```

Comment:
We cant put `REQUIRE`s into the `env` like `requres_errors (false, false)`, but it hides where a violation happens.